From 05ca1167c619c6060787d3128de6bab787f39a31 Mon Sep 17 00:00:00 2001
From: Shreeya Patel <spatel@ciq.com>
Date: Fri, 30 Jan 2026 10:37:54 +0000
Subject: [PATCH] KVM: arm64: Tear down vGIC on failed vCPU creation

jira VULN-67833
cve CVE-2025-37849
commit-author Will Deacon <will@kernel.org>
commit 250f25367b58d8c65a1b060a2dda037eea09a672

If kvm_arch_vcpu_create() fails to share the vCPU page with the
hypervisor, we propagate the error back to the ioctl but leave the
vGIC vCPU data initialised. Note only does this leak the corresponding
memory when the vCPU is destroyed but it can also lead to use-after-free
if the redistributor device handling tries to walk into the vCPU.

Add the missing cleanup to kvm_arch_vcpu_create(), ensuring that the
vGIC vCPU structures are destroyed on error.

	Cc: <stable@vger.kernel.org>
	Cc: Marc Zyngier <maz@kernel.org>
	Cc: Oliver Upton <oliver.upton@linux.dev>
	Cc: Quentin Perret <qperret@google.com>
	Signed-off-by: Will Deacon <will@kernel.org>
	Reviewed-by: Marc Zyngier <maz@kernel.org>
Link: https://lore.kernel.org/r/20250314133409.9123-1-will@kernel.org
	Signed-off-by: Oliver Upton <oliver.upton@linux.dev>
(cherry picked from commit 250f25367b58d8c65a1b060a2dda037eea09a672)
	Signed-off-by: Shreeya Patel <spatel@ciq.com>
---
 arch/arm64/kvm/arm.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 917086be5c6b..c6548ba40024 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -351,7 +351,11 @@ int kvm_arch_vcpu_create(struct kvm_vcpu *vcpu)
 	if (err)
 		return err;
 
-	return kvm_share_hyp(vcpu, vcpu + 1);
+	err = kvm_share_hyp(vcpu, vcpu + 1);
+	if (err)
+		kvm_vgic_vcpu_destroy(vcpu);
+
+	return err;
 }
 
 void kvm_arch_vcpu_postcreate(struct kvm_vcpu *vcpu)
-- 
2.47.3

